Epic 8080 : WNPRC Purchasing Module - Initial Setup

labkey-bpatel opened the pull request:
#### Rationale
Initial implementation of WNPRC Purchasing workflow

#### Related Pull Requests
https://github.com/LabKey/wnprc-modules/pull/46
https://github.com/LabKey/platform/pull/1746
https://github.com/LabKey/distributions/pull/116

#### Changes
- EHR Purchasing module files and setup
- Addition of ehr_purchasing schema, tables, metadata

labkey-bpatel committed:
EHR Purchasing Module
- Upgrade scripts with ehr_purchasing schema and tables with LSID col to make tables extensible
- Metadata
- Purchasing Admin role
- Update ContainerListener, override containerDeleted()

labkey-bpatel committed:
Make ehr_purchasing tables extensible. Fix upgrade script.

labkey-bpatel committed:
Add 'controlSubstance' to lineItems table

labkey-bpatel committed:
Add metadata, some cleanup around updating table names and IDX and FK names.

labkey-bpatel committed:
Make UQ cols required.

labkey-jeckels commented on EHR_Purchasing/resources/queries/ehr_purchasing/purchasingRequests.query.xml:
```diff
@@ -0,0 +1,34 @@
+                        <fk>
+                            <fkDbSchema>ehr_purchasing</fkDbSchema>
+                            <fkTable>vendor</fkTable>
+                            <fkColumnName>rowId</fkColumnName>
```
You can omit the fkColumnName and it will default to using the PK of the target table. Similarly, it's typically best to set the titleColumn on the target table instead of specifying it as the fkDisplayColumnName every time it's referenced.

labkey-bpatel replied:
good points, done!

labkey-jeckels commented on EHR_Purchasing/resources/schemas/dbscripts/postgresql/ehr_purchasing-20.000-20.001.sql:
```diff
@@ -0,0 +1,176 @@
+    vendorId         int,
+    account          varchar(200),
+    shippingInfoId   int,
+    justification    varchar(1000),
```
Let's make these just "varchar" instead of varchar(1000). There's no harm in not having a cap.

labkey-bpatel replied:
done

labkey-jeckels commented on EHR_Purchasing/resources/schemas/dbscripts/postgresql/ehr_purchasing-20.000-20.001.sql:
```diff
@@ -0,0 +1,176 @@
+    rowId      serial,
+    requestId  entityid,
+    item       varchar(500),
+    itemUnitId   int,
```
This and itemStatusId should have a FK and index, and probably be NOT NULL

labkey-bpatel replied:
My initial thought for not having a hard FK and index on these columns since they are "low cardinality" columns, but since we will be doing a join, having a hard FK (and index) makes sense - so its added now.

labkey-jeckels commented on EHR_Purchasing/src/org/labkey/ehr_purchasing/EHR_PurchasingContainerListener.java:
```diff
@@ -0,0 +1,95 @@
+        try (DbScope.Transaction transaction = scope.ensureTransaction())
+        {
+            TableInfo vendorTable = EHR_PurchasingSchema.getInstance().getVendorTable();
+            if (vendorTable.getTableType() == DatabaseTableType.TABLE)
```
Why this check? Shouldn't it always be a table in the DB?

labkey-bpatel replied:
removed

labkey-bpatel committed:
Code review changes:
- Metadata updates - omit fkColumnName where applicable, remove fkDisplayColumnName and set titleColumn instead
- Column updates - change to varchar, add FK and IDX to itemUnitId and itemStatusId
- Other minor cleanup

labkey-bpatel committed:
Include rowId for userAccountAssociations table. Re-format code for readability.

labkey-bpatel committed:
Update schema xml
